### PR TITLE
Better Chunk Store

### PR DESCRIPTION
**Pull request by Karlatemp**
Optimized the chunk store, now he will take up less space.

1 bool store in 1 byte  ->  8 bool store in 1 byte.

### CODE REVIEW

**Comment by nossr50** [comment]
I appreciate that you went through the effort to do this, before I'd be willing to review and test this branch I'd like to request more thorough documentation in your code. The chunk store is a mess and already lacks comments, I don't want to make it a bigger headache to maintain.

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 1]
```diff
@@ -47,51 +76,75 @@ public int getChunkZ() {
...
+        if (y < 0 || y >= worldHeight) return true;
+        return ((x | z) & ~0xF) != 0;
+    }
+
     @Override
     public boolean isTrue(int x, int y, int z) {
-        return store[x][z][y];
+        if (isInvalid(x, y, z)) return false;
+        int status = data.get((y * PAGE_SIZE) + (x >> 1));
+        return ((status >>> (z + (((x & 1) == 1) ? 16 : 0))) & 1) != 0;
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: Please include comments inside the method explaining the operations you are taking in this method, especially the bit shifting.

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 2]
```diff
@@ -47,51 +76,75 @@ public int getChunkZ() {
...
+        do {
+            source = data.get(point);
+            int target = source;
+            if (val) {
+                target |= 1 << bit;
+            } else {
+                target &= ~(1 << bit);
+            }
+            if (data.compareAndSet(point, source, target)) break;
+        } while (true);
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: Please include comments inside this method explaining the operations you are taking, especially the bit shifting.

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 3]
```diff
@@ -47,51 +76,75 @@ public int getChunkZ() {
...
-            for (int z = 0; z < 16; z++) {
-                for (int y = 0; y < store[0][0].length; y++) {
-                    if (store[x][z][y]) {
-                        return false;
-                    }
-                }
-            }
+        int end = data.length();
+        while (end-- > 0) {
+            if (data.get(end) != 0) return false;
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: I would appreciate comments explaining how this determines isEmpty()

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 4]
```diff
@@ -47,51 +76,75 @@ public int getChunkZ() {
...
+            while (end-- > 0) {
+                to.set(end, data.get(end));
+            }
+        } else {
+            int wh = worldHeight;
+            for (int x = 0; x < 16; x++) {
+                for (int z = 0; z < 16; z++) {
+                    for (int y = 0; y < wh; y++) {
+                        set(x, y, z, otherStore.isTrue(x, y, z));
+                    }
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: Please explain the operations you are taking here, especially the thoughts behind the nested for loops and the while loop.

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 5]
```diff
@@ -102,7 +155,11 @@ private void writeObject(ObjectOutputStream out) throws IOException {
         out.writeLong(worldUid.getMostSignificantBits());
         out.writeInt(cx);
         out.writeInt(cz);
-        out.writeObject(store);
+        int len;
+        out.writeInt(len = data.length());
+        for (int i = 0; i < len; i++) {
+            out.writeInt(data.get(i));
+        }
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: I would like comments here too even though its easy to infer what is going on

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 6]
```diff
@@ -122,7 +179,25 @@ private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundE
...
+            int wh;
+            this.data = new AtomicIntegerArray((wh = worldHeight = st[0][0].length) * PAGE_SIZE);
+            for (int x = 0; x < 16; x++) {
+                for (int z = 0; z < 16; z++) {
+                    for (int y = 0; y < wh; y++) {
+                        set(x, y, z, st[x][z][y]);
+                    }
+                }
+            }
+        }
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: Please add comments to explain the operations you are going through here

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 7]
```diff
@@ -131,17 +206,13 @@ private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundE
...
-                    catch (Exception e) { e.printStackTrace(); }
-                }
-            }
+        AtomicIntegerArray s = this.data;
+        AtomicIntegerArray d = this.data = new AtomicIntegerArray(PAGE_SIZE * (
+                worldHeight = Bukkit.getWorld(worldUid).getMaxHeight()
+        ));
+        int end = Math.max(s.length(), d.length());
+        while (end-- > 0) {
+            d.set(end, s.get(end));
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: Please add comments explaining the operations you are going through

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 8]
```diff
@@ -131,17 +206,13 @@ private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundE
...
-            for (int z = 0; z < 16; z++) {
-                for (int y = 0; y < store[0][0].length; y++) {
-                    try {
-                        store[x][z][y] = temp[x][y][z];
-                    }
-                    catch (Exception e) { e.printStackTrace(); }
-                }
-            }
+        AtomicIntegerArray s = this.data;
+        AtomicIntegerArray d = this.data = new AtomicIntegerArray(PAGE_SIZE * (
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: Please avoid one letter variables, try to follow Java Code Conventions.

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 9]
```diff
@@ -122,7 +179,25 @@ private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundE
...
+        if (fileVersionNumber == CURRENT_VERSION) {
+            int size;
+            AtomicIntegerArray data = this.data = new AtomicIntegerArray(size = in.readInt());
+            worldHeight = size / PAGE_SIZE;
+            for (int i = 0; i < size; i++) {
+                data.set(i, in.readInt());
+            }
+        } else {
+            boolean[][][] st = (boolean[][][]) in.readObject();
+            int wh;
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: Please avoid variable names of this nature, I would prefer more verbose names.

**Comment by Karlatemp** [comment]
Yes, comments have been added, but I have already written a comment about `bits shifting` in the field `AtomicIntegerArray data`

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 10]
```diff
@@ -180,16 +211,21 @@ private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundE
...
+                throw new InvalidObjectException("Illegal data length");
             worldHeight = size / PAGE_SIZE;
             for (int i = 0; i < size; i++) {
                 data.set(i, in.readInt());
             }
         } else {
+            // File version 7
             boolean[][][] st = (boolean[][][]) in.readObject();
-            int wh;
+            int wh; // world height
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: I'd prefer a variable name like worldHeight

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 11]
```diff
@@ -206,13 +242,15 @@ private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundE
...
 
     private void fixArray() {
-        AtomicIntegerArray s = this.data;
-        AtomicIntegerArray d = this.data = new AtomicIntegerArray(PAGE_SIZE * (
+        AtomicIntegerArray old = this.data;
+        AtomicIntegerArray current = this.data = new AtomicIntegerArray(PAGE_SIZE * (
                 worldHeight = Bukkit.getWorld(worldUid).getMaxHeight()
         ));
-        int end = Math.max(s.length(), d.length());
+        // To ensure that no cross-border
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: define what you mean by cross border, it's not obvious

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 12]
```diff
@@ -118,28 +133,43 @@ public void setFalse(int x, int y, int z) {
...
+            // into: ---- ---- ---- ----
+            int end = Math.min(from.length(), into.length());
+            // same as for(int i = end.length()-1; i >= 0; i--)
             while (end-- > 0) {
-                to.set(end, data.get(end));
+                into.set(end, from.get(end));
             }
         } else {
-            int wh = worldHeight;
+            // Copy values with legacy method.
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: I'd prefer a comment explaining the process of this nested loop, I understand you didn't write it so I could let it slide.

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 13]
```diff
@@ -118,28 +133,43 @@ public void setFalse(int x, int y, int z) {
...
+            AtomicIntegerArray into = this.data = new AtomicIntegerArray(
+                    worldHeight * PAGE_SIZE
+            );
+            // To ensure that no cross-border
+            // Like:
+            // from: ---- ---- ---- ----
+            // into: ---- ---- ---- ---- ---- ---- ---- ----
+
+            // from: ---- ---- ---- ---- ---- ---- ---- ----
+            // into: ---- ---- ---- ----
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: I'm not sure I understand what this is referring to

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 14]
```diff
@@ -94,14 +105,18 @@ private void set(int x, int y, int z, boolean val) {
         int point = (y * PAGE_SIZE) + (x >> 1);
         int bit = z + (((x & 1) == 1) ? 16 : 0);
         int source;
+        // An explanation of the operation of shift is on lines 26 to 42
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: Line numbers change when the file is edited, I'd prefer a comment that didn't need to be regularly updated

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 15]
```diff
@@ -67,25 +71,32 @@ public void setDirty(boolean dirty) {
...
         if (y < 0 || y >= worldHeight) return true;
-        return ((x | z) & ~0xF) != 0;
+        return ((x | z) & ~0xF) != 0; // < 0 || > 15
     }
 
     @Override
+    @Contract(pure = true)
     public boolean isTrue(int x, int y, int z) {
         if (isInvalid(x, y, z)) return false;
+        // An explanation of the operation of shift is on lines 26 to 42
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: Line numbers change when the file is edited, I'd prefer a comment that didn't need to be regularly updated

File: src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java [hunk 16]
```diff
@@ -40,7 +43,7 @@
      * boolean isOk = ((code >>> offset) & 1) == 1;
      * </pre>
      */
-    public AtomicIntegerArray data;
+    public AtomicIntegerArray data; // should it private or protected?
```
Comment (nossr50) [inline src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java]: depends on whether or not you want classes in the same package to be able to access it

**Comment by nossr50** [comment]
No response from the author, and on top of that @t00thpick1 is rewriting this code.

**Comment by t00thpick1** [comment]
Similar improvements merged in 01f31e76f57e8be8bbaa867bf6396bbaf752f792

**Pull request closed without merging**